# **LAB 1: Fundamentals**

## **General Goal**
The goal of this project is to introduce the functionalities and a general workorder when using Image Processing 
and Toolboxes in Numpy and OpenCV

Check up that all works ok ! :D

In [1]:
import cv2 as cv
import numpy as np
 
print("OpenCV:", cv.__version__)
img = np.zeros((120, 400, 3), dtype=np.uint8)
cv.putText(img, "OpenCV OK", (10, 80), cv.FONT_HERSHEY_SIMPLEX, 2, (255,255,255), 3)
# If you installed a non-headless build, you can display a window:
# cv.imshow("hello", img); cv.waitKey(0)
# Always safe (headless or not): save to file
cv.imwrite("hello.png", img)

OpenCV: 4.12.0


True

## **Exercise 1 - Reading Image Types**

1. Open the image Original.png in Python using using the CV2.imread function, list all the image
properties in a commented section in your script.

In [2]:
# Load the image
image = cv.imread("Images_Lab_1\Original.png")

# Check if the image loaded successfully
if image is None:
    print("Error: Could not load image. Make sure Original.png is in the working directory.")
else:
    # Display image properties (printed)
    print("Image Properties:")
    print(f" - Shape (Height, Width, Channels): {image.shape}")
    print(f" - Image Height: {image.shape[0]}")
    print(f" - Image Width: {image.shape[1]}")
    print(f" - Number of Channels: {image.shape[2] if len(image.shape) == 3 else 1}")
    print(f" - Data Type: {image.dtype}")
    print(f" - Total Pixels: {image.size}")

"""
IMAGE PROPERTIES (Explained)
----------------------------
image.shape
    - Returns a tuple: (height, width, channels)
      Example: (720, 1280, 3)

image.shape[0]
    - Height in pixels.

image.shape[1]
    - Width in pixels.

image.shape[2]
    - Number of color channels (e.g., 3 for BGR images).

image.dtype
    - Data type of the image array, typically uint8.

image.size
    - Total number of pixel values = height * width * channels
"""


Image Properties:
 - Shape (Height, Width, Channels): (540, 1050, 3)
 - Image Height: 540
 - Image Width: 1050
 - Number of Channels: 3
 - Data Type: uint8
 - Total Pixels: 1701000


<>:2: SyntaxWarning: invalid escape sequence '\O'
<>:2: SyntaxWarning: invalid escape sequence '\O'
C:\Users\pablo\AppData\Local\Temp\ipykernel_8924\3451966476.py:2: SyntaxWarning: invalid escape sequence '\O'
  image = cv.imread("Images_Lab_1\Original.png")


'\nIMAGE PROPERTIES (Explained)\n----------------------------\nimage.shape\n    - Returns a tuple: (height, width, channels)\n      Example: (720, 1280, 3)\n\nimage.shape[0]\n    - Height in pixels.\n\nimage.shape[1]\n    - Width in pixels.\n\nimage.shape[2]\n    - Number of color channels (e.g., 3 for BGR images).\n\nimage.dtype\n    - Data type of the image array, typically uint8.\n\nimage.size\n    - Total number of pixel values = height * width * channels\n'

2. There is a property in Original.png that has been automatically added by the CV2.imread function. Find
the property using the CV2.imread function. What is the property and how does it affect the information
in the image?

A 3-channel BGR color representation (forced by cv2.IMREAD_COLOR).
What actually happens internally?

- If the PNG was grayscale (1 channel) → OpenCV expands it to 3 channels using BGR.

- If the PNG had RGBA (4 channels) → OpenCV drops the alpha channel and returns 3-channel BGR.

- If it had RGB → OpenCV keeps 3 channels but converts to BGR ordering.

## **Exercise 2 – Finding Color values**
In the previous exercise, you learned how to detect how many color channels an image has. Now it is time to 
extract specific RGB values in images. Find the RGB values in image colormap.jpg for the four positions indicated below, you will need to know what 
pixel to look for in the image. For this it can be useful to use paint as help: 

In [3]:
im2 = cv.imread("Images_Lab_1\colormap.jpg")

positions = [(390, 20), (230, 130), (400, 300), (130, 340)]


for i, (x, y) in enumerate(positions):
    print(f"{i+1}. Pixel in position ({x},{y}), has a RGB value = {im2[x,y,:]}")


1. Pixel in position (390,20), has a RGB value = [255 255 255]
2. Pixel in position (230,130), has a RGB value = [ 52 204  51]
3. Pixel in position (400,300), has a RGB value = [  0  71 227]
4. Pixel in position (130,340), has a RGB value = [252 153  49]


<>:1: SyntaxWarning: invalid escape sequence '\c'
<>:1: SyntaxWarning: invalid escape sequence '\c'
C:\Users\pablo\AppData\Local\Temp\ipykernel_8924\606044286.py:1: SyntaxWarning: invalid escape sequence '\c'
  im2 = cv.imread("Images_Lab_1\colormap.jpg")


What are the RGB values for the four points? Leave all your answers in the comments box inside the script.

1. Pixel in position (390,20), has a RGB value = [255 255 255]
2. Pixel in position (230,130), has a RGB value = [ 52 204  51]
3. Pixel in position (400,300), has a RGB value = [  0  71 227]
4. Pixel in position (130,340), has a RGB value = [252 153  49]

## **Exercise 3 – Binarization, Image stacking and trackbars**

When working with images, displaying them side by side after making manipulations to them can be useful to 
follow what is going on when changing values in filters. Create a copy of the image “piece03.png” using the five methods below 

- cv.THRESH_BINARY
- cv.THRESH_BINARY_INV
- cv.THRESH_TRUNC
- cv.THRESH_TOZERO
- cv.THRESH_TOZERO_INV

Now you should have five different binarized versions of “piece03.png”. You can view this as having five different and independent matrices containing image data for each method: 

Now use a list to merge these images together and the NumPy stack function to display them vertically and 
horizontally. Think about how imshow reads the picture after this operation, what happens?  

In [4]:
# ===========================[  LOAD IMAGE  ]=====================================
img = cv.imread("Images_Lab_1\piece03.png", cv.IMREAD_GRAYSCALE)
if img is None:
    raise FileNotFoundError("piece03.png not found.")

max_value = 255


# ===========================[  TRACKBAR CALLBACK  ]==============================
def nothing(x):
    pass


# ===========================[  WINDOW + TRACKBARS  ]=============================
cv.namedWindow("Threshold Tuner")

cv.createTrackbar("THRESH", "Threshold Tuner", 128, 255, nothing)


while True:
    # Read trackbar position
    t = cv.getTrackbarPos("THRESH", "Threshold Tuner")

    # Apply all five thresholding methods
    _, th_binary     = cv.threshold(img, t, max_value, cv.THRESH_BINARY)
    _, th_binary_inv = cv.threshold(img, t, max_value, cv.THRESH_BINARY_INV)
    _, th_trunc      = cv.threshold(img, t, max_value, cv.THRESH_TRUNC)
    _, th_tozero     = cv.threshold(img, t, max_value, cv.THRESH_TOZERO)
    _, th_tozero_inv = cv.threshold(img, t, max_value, cv.THRESH_TOZERO_INV)

    # Put in a visual grid (convert to BGR for display)
    def lab(text, im):
        im = cv.cvtColor(im, cv.COLOR_GRAY2BGR)
        cv.putText(im, text, (10, 25), cv.FONT_HERSHEY_SIMPLEX,
                   0.7, (255, 255, 255), 2, cv.LINE_AA)
        return im

    images = [
        lab("BINARY", th_binary),
        lab("BINARY_INV", th_binary_inv),
        lab("TRUNC", th_trunc),
        lab("TOZERO", th_tozero),
        lab("TOZERO_INV", th_tozero_inv)
    ]

    # Pad images to make grid look nicer
    def pad(im):
        return cv.copyMakeBorder(im, 10, 10, 10, 10,
                                 cv.BORDER_CONSTANT, value=(30, 30, 30))

    images = [pad(im) for im in images]

    # 2x3 grid (last one empty)
    h, w = images[0].shape[:2]
    empty = np.zeros((h, w, 3), dtype=np.uint8)
    images.append(empty)

    row1 = np.hstack(images[:3])
    row2 = np.hstack(images[3:6])
    grid = np.vstack((row1, row2))

    display = cv.resize(grid, None, fx=0.1, fy=0.1)  # scale down to 10%
    cv.imshow("Threshold Tuner", display)


    # Quit on ESC
    if cv.waitKey(10) & 0xFF == 27:
        break

cv.destroyAllWindows()


<>:2: SyntaxWarning: invalid escape sequence '\p'
<>:2: SyntaxWarning: invalid escape sequence '\p'
C:\Users\pablo\AppData\Local\Temp\ipykernel_8924\908840032.py:2: SyntaxWarning: invalid escape sequence '\p'
  img = cv.imread("Images_Lab_1\piece03.png", cv.IMREAD_GRAYSCALE)


The three pieces (bright objects) appear most clearly when the threshold 
    value is roughly in the range **80–150**, depending on the image.
    Below ~80 the threshold is too low and too much background is included.
    Above ~150 the threshold removes parts of the pieces or loses smaller ones.
    So a useful range where all threshold methods still show the three objects
    is approximately **[80, 150]**.

A2: Because an 8-bit grayscale image only contains intensity values from 
    **0 to 255**. Increasing the trackbar range (e.g., 0–1000) gives no real 
    improvement—threshold values above 255 have **no effect** on the image.
    Each pixel is stored as 1 byte, so any value outside 0–255 is meaningless.


## **Exercise 4 – Extracting Color Channels**
Read the image colormap.jpg and extract its color channels separately in:

In [7]:
import cv2 as cv
import numpy as np

grayscale = cv.imread("Images_Lab_1\\colormap.jpg")

# Extract channels
B = grayscale[:, :, 0]
G = grayscale[:, :, 1]
R = grayscale[:, :, 2]

# ---------------------------
# 1) Grayscale channel row
# ---------------------------
gray_images = (B, G, R)
gray_row = np.hstack(gray_images)

# Convert the grayscale row (2D) to BGR (3D) so it matches the colored row dimensions
gray_row_bgr = cv.cvtColor(gray_row, cv.COLOR_GRAY2BGR)

# ---------------------------
# 2) Colored channel row
# ---------------------------
# Create blank channels for merging
zeros = np.zeros_like(B)

blue_colored  = cv.merge((B, zeros, zeros))   # Blue channel emphasized
green_colored = cv.merge((zeros, G, zeros))   # Green channel emphasized
red_colored   = cv.merge((zeros, zeros, R))   # Red channel emphasized

colored_images = (blue_colored, green_colored, red_colored)
color_row = np.hstack(colored_images)

# ---------------------------
# 3) Stack rows vertically
# ---------------------------
final = np.vstack((gray_row_bgr, color_row))

cv.imshow("Grayscale + Colored Channels", final)
cv.waitKey(0)
cv.destroyAllWindows()


## **Exercise 5 – Preparing image for edge detection** 

To detect objects in an image, a common approach is to find the contours of the objects against the image background.  

The contours are referred to as edges in image processing, and edges are registered when there is a sudden shift 
in pixel value between neighbouring pixels.   

**The objective for exercise 5 is to find get only 4 objects outputted from cv2.findContours when analysing piece05.png.**

In https://www.geeksforgeeks.org/find-and-draw-contours-using-opencv-python/ you can see an example of this operation as well as some sample code. However, the images used in that example are not representative to a real world image which usually contains colour information in RGB or grayscale format, so piece05.png needs further processing before the function cv2.findContours can be used. **In essence the result needs to be 4 when running the line print("Number of Contours found = " + str(len(contours)))** 

To get there, several pre-processing steps needs to be performed to an image to get better perform edge detection, you can see examples of this in the link as well. 

Perform the following steps to the image piece05.png by prompting a LLM of your selection.  The preprocessing steps follow a standard structure that any LLM generates. 

- Explain the different steps and why they are necessary inside of your code by writing comments where the filters are used.   In a certain step, it uses canny edge detection and sometimes binarization.  

- What filter is the best in what circumstance, is there something you can do when acquiring the image that will increase the accuracy of the image pre-processing? Write your answer in the bottom of your script.

In [ ]:
import cv2 as cv
import numpy as np

# =========================
# LOAD IMAGE
# =========================
img = cv.imread("Images_Lab_1/piece05.png")
if img is None:
    raise FileNotFoundError("piece05.png not found.")

# Convert to grayscale (contour detection works best on single-channel images)
gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY)

# =========================
# CREATE WINDOW + TRACKBARS
# =========================
cv.namedWindow("Contour Detection Tuner")

# Trackbars for Gaussian blur kernel size (must be odd)
cv.createTrackbar("Blur", "Contour Detection Tuner", 1, 20, lambda x: None)
# Trackbars for Thresholding
cv.createTrackbar("Thresh", "Contour Detection Tuner", 128, 255, lambda x: None)
cv.createTrackbar("MaxVal", "Contour Detection Tuner", 255, 255, lambda x: None)
# Trackbars for Canny edges
cv.createTrackbar("Canny Low", "Contour Detection Tuner", 50, 500, lambda x: None)
cv.createTrackbar("Canny High", "Contour Detection Tuner", 150, 500, lambda x: None)
# Trackbars for Morphology (optional to close gaps)
cv.createTrackbar("Morph Kernel", "Contour Detection Tuner", 1, 20, lambda x: None)


while True:
    # =========================
    # READ TRACKBAR VALUES
    # =========================
    ksize = cv.getTrackbarPos("Blur", "Contour Detection Tuner")
    if ksize % 2 == 0:  # Kernel size must be odd
        ksize += 1

    thresh_val = cv.getTrackbarPos("Thresh", "Contour Detection Tuner")
    max_val = cv.getTrackbarPos("MaxVal", "Contour Detection Tuner")
    canny_low = cv.getTrackbarPos("Canny Low", "Contour Detection Tuner")
    canny_high = cv.getTrackbarPos("Canny High", "Contour Detection Tuner")

    morph_k = cv.getTrackbarPos("Morph Kernel", "Contour Detection Tuner")
    if morph_k % 2 == 0:
        morph_k += 1

    # =========================
    # STEP 1: SMOOTHING
    # =========================
    # Gaussian blur reduces small noise and intensity variations
    blurred = cv.GaussianBlur(gray, (ksize, ksize), 0)

    # =========================
    # STEP 2: THRESHOLDING
    # =========================
    # Converts image to binary to separate objects from background
    _, binary = cv.threshold(blurred, thresh_val, max_val, cv.THRESH_BINARY)

    # =========================
    # STEP 3: CANNY
    # =========================
    # Detect edges by finding sudden intensity changes
    edges = cv.Canny(blurred, canny_low, canny_high)

    # =========================
    # STEP 4: MORPH CLOSE
    # =========================
    # Closing gaps in edges to get full contours
    kernel = cv.getStructuringElement(cv.MORPH_ELLIPSE, (morph_k, morph_k))
    edges_closed = cv.morphologyEx(edges, cv.MORPH_CLOSE, kernel)

    # =========================
    # STEP 5: FIND CONTOURS
    # =========================
    contours, _ = cv.findContours(edges_closed, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)

    cont_img = img.copy()
    cv.drawContours(cont_img, contours, -1, (0, 255, 0), 2)

    # Add contour count text
    cv.putText(
        cont_img,
        f"Contours: {len(contours)}",
        (150, 150),
        cv.FONT_HERSHEY_SIMPLEX,
        5.0,
        (0, 0, 255),
        10
    )

    # =========================
    # DISPLAY ALL STEPS IN A ROW
    # =========================
    disp_h = 250

    def r(img):
        return cv.resize(img, (int(img.shape[1] * disp_h / img.shape[0]), disp_h))

    gray_bgr    = cv.cvtColor(r(gray), cv.COLOR_GRAY2BGR)
    blurred_bgr = cv.cvtColor(r(blurred), cv.COLOR_GRAY2BGR)
    binary_bgr  = cv.cvtColor(r(binary), cv.COLOR_GRAY2BGR)
    edges_bgr   = cv.cvtColor(r(edges_closed), cv.COLOR_GRAY2BGR)
    cont_small  = r(cont_img)

    row = np.hstack([
        gray_bgr,
        blurred_bgr,
        binary_bgr,
        edges_bgr,
        cont_small
    ])

    cv.imshow("Pipeline Overview", row)

    # Main window image
    scale = 0.3
    display = cv.resize(cont_img, None, fx=scale, fy=scale)
    cv.imshow("Contour Detection Tuner", display)

    # Quit on ESC
    if cv.waitKey(10) & 0xFF == 27:
        break

cv.destroyAllWindows()


📝 Comments on Filters and Image Acquisition
1) Gaussian Blur

Reduces noise to avoid false edges.

Works well for images with grainy or textured backgrounds.

2) Thresholding / Binarization

Separates foreground objects from background.

Effective when objects have clear intensity differences from the background.

Adaptive thresholding helps when lighting is uneven.

3) Canny Edge Detection

Detects edges using intensity gradients.

Sensitive to noise → smoothing first is important.

Best when accurate shape boundaries are required.

4) Morphological Operations

Closing (dilation → erosion) connects broken edges.

Useful when contours have small gaps or noise.

5) Best Filter for Each Situation

High contrast objects: simple THRESH_BINARY.

Detailed objects: Gaussian blur → Canny edges.

Noisy or uneven lighting: Gaussian blur → Adaptive threshold → Morphological closing.

6) Image Acquisition Tips

Maintain uniform lighting and high contrast between object and background.

Use fast shutter speeds to reduce motion blur.

Reduce camera noise (low ISO, clean lens/sensor).

Ensure objects are fully in frame and well separated.

## **Exercise 6 – Finding Position**

 Using OpenCV, find the centre position in pixels of the object inside the colour image star.png by 

- Using a canny filter    
- Using Binarization filter

Leave all your answers in the comments box inside the script clearly diversifying which coordinates you get 
when using the different filters. 

In [2]:
import cv2 as cv
import numpy as np

# Load image
img = cv.imread("Images_Lab_1/star.png")
gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY)

# Canny edge detection
edges = cv.Canny(gray, 50, 150)

# Find contours
contours, _ = cv.findContours(edges, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)

# Take the largest contour (the star)
cnt = max(contours, key=cv.contourArea)

# Compute centroid using image moments
M = cv.moments(cnt)
cx = int(M['m10'] / M['m00'])
cy = int(M['m01'] / M['m00'])

print("Canny centroid =", (cx, cy))

# Draw centroid
cv.circle(img, (cx, cy), 6, (0, 255, 0), -1)
cv.imshow("Canny Center", img)
cv.waitKey(0)
cv.destroyAllWindows()


Canny centroid = (88, 113)


In [6]:
# Load image (try the Images_Lab_1 path used elsewhere, then fallback to current dir)
img = cv.imread("Images_Lab_1/star.png")
if img is None:
	img = cv.imread("star.png")
if img is None:
	raise FileNotFoundError("star.png not found. Place it in 'Images_Lab_1/' or the notebook working directory.")

# Convert to grayscale (safe because img is confirmed loaded)
gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY)

# Inverted threshold so the star is white and background is black
_, binary = cv.threshold(gray, 128, 255, cv.THRESH_BINARY_INV)

# Find contours
contours, _ = cv.findContours(binary, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)

if len(contours) == 0:
	raise RuntimeError("No contours found in the binary image. Check thresholding or image path.")

# Largest contour = star
cnt = max(contours, key=cv.contourArea)

# Compute centroid (protect against zero area)
M = cv.moments(cnt)
if M['m00'] == 0:
	raise RuntimeError("Contour has zero area, cannot compute centroid.")
cx = int(M['m10'] / M['m00'])
cy = int(M['m01'] / M['m00'])

print("Binary centroid =", (cx, cy))

# Draw centroid on a copy to avoid modifying original unexpectedly
out = img.copy()
cv.circle(out, (cx, cy), 6, (0, 255, 0), -1)
cv.imshow("Binary Center", out)
cv.imshow("Binary Mask", binary)
cv.waitKey(0)
cv.destroyAllWindows()


Binary centroid = (88, 113)


## **Exercise 7 – Finding Orientation**  
Using OpenCV, Find the orientation in degrees of the object inside the color image tree.png by 

- Using a canny filter  
- Using binarization filter

In [13]:
import cv2 as cv
import numpy as np
import math

# Load image with fallbacks to common paths to avoid None image
img = cv.imread("Images_Lab_1/tree.png")
if img is None:
	img = cv.imread("Images_Lab_1/tree.png")
if img is None:
	img = cv.imread("Images_Lab_1\\tree.png")
if img is None:
	raise FileNotFoundError("tree.png not found. Place it in the notebook working directory or 'Images_Lab_1/'.")

# Convert to grayscale (safe now since img is loaded)
gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY)

# --- Canny edges ---
edges = cv.Canny(gray, 50, 150)

# Find contours
contours_info = cv.findContours(edges, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_NONE)
# Compatibility for OpenCV versions (findContours returns 2 or 3 values)
if len(contours_info) == 3:
	_, contours, _ = contours_info
else:
	contours, _ = contours_info

# Ensure we have at least one contour
if len(contours) == 0:
	raise RuntimeError("No contours found. Try adjusting Canny thresholds or check the input image.")

# Use the largest contour (tree)
cnt = max(contours, key=cv.contourArea)

# Convert contour to Nx2 array of points
data_pts = cnt.reshape(-1, 2).astype(np.float32)

# PCA
# cv.PCACompute2 requires a 'mean' argument (can be None to compute internally)
mean, eigenvectors, eigenvalues = cv.PCACompute2(data_pts, mean=None)

# Orientation angle (principal eigenvector)
vx, vy = eigenvectors[0]
angle = math.degrees(math.atan2(vy, vx))

print("Orientation using Canny =", angle, "degrees")

# Optional: draw axis
# mean may be shape (1,2), flatten it to get coordinates
center_coords = tuple(int(x) for x in mean.ravel())
p2 = (int(center_coords[0] + vx * 100), int(center_coords[1] + vy * 100))
cv.circle(img, center_coords, 5, (0, 0, 255), -1)
cv.line(img, center_coords, p2, (0, 255, 0), 3)

cv.imshow("Orientation (Canny)", img)
cv.waitKey(0)
cv.destroyAllWindows()


Orientation using Canny = 125.2736486368268 degrees


In [12]:
import cv2 as cv
import numpy as np
import math

# Load image (try common paths)
img = cv.imread("Images_Lab_1/tree.png")
if img is None:
	img = cv.imread("Images_Lab_1\\tree.png")
if img is None:
	img = cv.imread("tree.png")
if img is None:
	raise FileNotFoundError("tree.png not found. Place it in 'Images_Lab_1/' or the notebook working directory.")

# Convert to grayscale
gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY)

# Threshold (invert if tree is dark)
_, binary = cv.threshold(gray, 128, 255, cv.THRESH_BINARY_INV)

# Find contours (compatibility for different OpenCV returns)
contours_info = cv.findContours(binary, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_NONE)
if len(contours_info) == 3:
	_, contours, _ = contours_info
else:
	contours, _ = contours_info

if len(contours) == 0:
	raise RuntimeError("No contours found. Check thresholding or image path.")

# Largest contour = tree
cnt = max(contours, key=cv.contourArea)

# Prepare points for PCA
data_pts = cnt.reshape(-1, 2).astype(np.float32)

# PCA
# Some OpenCV builds require an explicit 'mean' argument for PCACompute2.
# Provide None to let OpenCV compute the mean internally.
# This call returns mean, eigenvectors, eigenvalues (or similar depending on version).
mean, eigenvectors, eigenvalues = cv.PCACompute2(data_pts, None)

# Orientation (principal eigenvector)
vx, vy = eigenvectors[0]
angle = math.degrees(math.atan2(vy, vx))

print("Orientation using Binarization =", angle, "degrees")

# Optional: draw axis
# Ensure mean is flattened to (x,y) ints
center = tuple(int(x) for x in mean.ravel())
p2 = (int(center[0] + vx * 100), int(center[1] + vy * 100))
cv.circle(img, center, 5, (0, 0, 255), -1)
cv.line(img, center, p2, (0, 255, 0), 3)

cv.imshow("Orientation (Binary)", img)
cv.imshow("Binary Mask", binary)
cv.waitKey(0)
cv.destroyAllWindows()


Orientation using Binarization = 125.68969276065359 degrees
